tesseract_common/include/tesseract_common/plugin_loader.h

```diff
@@ -90,6 +91,19 @@ class PluginLoader
    */
   inline bool isPluginAvailable(const std::string& plugin_name) const;
 
+  /**
+   * @brief Get the available plugins under the provided section
+   * @param section The section name to get all available plugins
+   * @return A list of available plugins under the provided section
+   */
+  inline std::vector<std::string> getAvailablePlugins(const std::string& section) const;
+
+  /**
+   * @brief Get the available sections within the provided search libraries
+   * @return A list of available sections
+   */
+  inline std::vector<std::string> getAvailableSections(bool include_hidden = false) const;
```

Comment:
I think this method has its uses, but I think a more common use case is to search for plugins by interface type. It's much easier to programmatically search for plugins of a specific type than to sift through a bunch of nondescript strings in a vector (i.e. `ContColl`) to find the section name (which isn't really exposed anywhere) for the plugin type I care about. 

I think my preferred approach would be something like `plugin_loader.getAvailablePlugins<tesseract_kinematics::InverseKinematicsPlugin>()`.  The the plugin loader could implement this by accessing a static variable in the plugin interface class with the section name. Since the section names have to be less than 8 bytes and aren't really descriptive, I think it would make sense to make it a private static variable in the class; users don't really ever need to know or see that section name.

Reply:
Currently that is accomplished by the `getAvailablePlugins(const std::string& section)`. Right now you can do the following.

``` c++
std::vector<std::string> plugins = plugin_loader.getAvailablePlugins("InvKin");
auto plugin = plugin_loader.instantiate<tesseract_kinematics::InverseKinematicsPlugin>(plugins[0]);
```

> The the plugin loader could implement this by accessing a static variable in the plugin interface class with the section name. Since the section names have to be less than 8 bytes and aren't really descriptive, I think it would make sense to make it a private static variable in the class; users don't really ever need to know or see that section name.

That would require that the plugin loader class is aware of each plugin type which I don't think we want to do. Am I missing something?

Reply:
> `std::vector<std::string> plugins = plugin_loader.getAvailablePlugins("InvKin");
auto plugin = plugin_loader.instantiate<tesseract_kinematics::InverseKinematicsPlugin>(plugins[0]);`

This is more or less what I *don't* want to do because the section names are pretty nondescript and not exposed to the user (currently at least). 

To do what I'm describing I think you would need to have a simple base class for plugins that forces them to provide a section name.

```c++
class Plugin
{
protected:
  std::string getSectionName() = 0;
}

class InvKinFactory : public Plugin
{
...
protected:
  std::string getSectionName() { return "InvKin"; }
  friend class PluginLoader;
}

class PluginLoader
{
...
<template typename PluginT>
std::vector<std::string> getAvailablePlugins()
{
  return class_loader_.getAvailablePlugins(..., T::getSectionName());
}
...
}
```

An easier solution would also be to make the section name a public static variable in the plugin class. Then you could do something like:
```c++
std::vector<std::string> plugins = plugin_loader.getAvailablePlugins(InvKinFactory::SECTION_NAME);
```
That would work fine too, it just seems cleaner to hide the section names from the user completely since they're only used internally

Reply:
I haven't thought through it completely, but the base class may not actually be required. They may just need a static variable and friend class declaration

Reply:
@marip8 Take a look at the latest commit adding the protected SECTION_NAME and added the templated method to `PluginLoader` which calls the `getAvailablePlugins` with the SECTION_NAME under the hood.